### PR TITLE
Introducing dark and light themes. Also, make it possible to change them on common settings

### PR DESCRIPTION
**Pull request by pdelfino**
This is a draft for issue #2043.

@aartaka suggested:

> @pdelfino, would you be interested is implementing theme settings, as we discussed via email (+dark-theme+ and +light-theme+ variables, Common Settings page with color samples and ability to change the theme). If not, I can do it sometime this week :)

I am not sure about the "color samples" on the Common Settings. Artyom, could you give me a hint on how to implement that?

Also, apart from the "color sample", is the rest of the PR fully implementing your suggestion?

Thanks!

### CODE REVIEW

**Comment by pdelfino** [comment]
In addition, I think you *did not* suggest this, Artyom... But since `(configure-slot` works permanents (since it changes `auto-config.lisp`) and requires re-start, I thought it would be cool to add some commands that change theme instantaneously and just for the current session.

0 - not sure if this is a good idea at all. Curious about your thoughts.
1 - not sure if the `user-interface.lisp` is the best place to put the commands
2 - it seems to work, except for the lower part of the U.I. (status area, icons, message area...). I do not fully understand the side effect behavior here. I will try to study it tomorrow with a fresh mind.

File: source/user-interface.lisp [hunk 1]
```diff
@@ -23,3 +23,13 @@
    (ps:ps (let ((element (ps:chain document (get-element-by-id (ps:lisp (user-interface:id button))))))
             (setf (ps:chain element text-content) (ps:lisp (user-interface:text button)))
             (setf (ps:chain element onclick) (ps:lisp (user-interface:action button)))))))
+
+(define-command set-dark-theme ()
+  "Set the THEME of BROWSER to be dark for this session. This does not change
+future sessions of Nyxt. To do that, see the Common Settings"
+  (Setf (Theme *Browser*) Theme::+Dark-Theme+))
```
Comment (aartaka) [inline source/user-interface.lisp:30]: Why this capitalization? 0_o

File: source/user-interface.lisp [hunk 2]
```diff
@@ -23,3 +23,13 @@
    (ps:ps (let ((element (ps:chain document (get-element-by-id (ps:lisp (user-interface:id button))))))
             (setf (ps:chain element text-content) (ps:lisp (user-interface:text button)))
             (setf (ps:chain element onclick) (ps:lisp (user-interface:action button)))))))
+
+(define-command set-dark-theme ()
+  "Set the THEME of BROWSER to be dark for this session. This does not change
```
Comment (aartaka) [inline source/user-interface.lisp:28]: Not sure you need to capitalise `browser` here -- it's not one of the arguments. Same for `theme`. You can use the backtick-quote notation, though :)

Comment (aartaka) [inline source/user-interface.lisp:28]: Make sure the first line of the docstring is strictly one sentence.

File: source/user-interface.lisp [hunk 3]
```diff
@@ -23,3 +23,13 @@
...
+future sessions of Nyxt. To do that, see the Common Settings"
+  (Setf (Theme *Browser*) Theme::+Dark-Theme+))
+
+(define-command set-light-theme ()
+  "Set the THEME of BROWSER to be light for this session. This does not change
```
Comment (aartaka) [inline source/user-interface.lisp:33]: Same. `theme` and `browser` are not to be capitalized and there should only be one sentence on the first docstring line.

**Review by aartaka — COMMENTED** [review]
All good, except for some stylistic moments! It's exactly as I imagined it looking!

**Regarding color snippets:** I was thinking of something along the lines of 
```
(spinneret:with-html-string
  (theme:with-theme (nyxt:theme nyxt:*browser*)
    (:span :style (format nil "background-color: ~a" theme:primary)
           theme:primary)))
```

This is just one color, so you'd need to repeat this generation for all of those, make sure no color is blending with the background (maybe a contrasting border?), and ensure the text is contrasting enough.

**Regarding commands**: I'm not sure we need those. What's the use-case? If one wants to enable dark theme for just one website/page, they'll use `dark-mode`. If it's the other way around -- they'll use `invert-color`. Maybe these commands don't belong to the patch. Mind opening an issue about this after we merge?

**Regarding `setf` not working as intended**: That's how it's expected to behave, as buffer/status-buffer/message-buffer style is set at the initialization time. If we change the theme at runtime, those are not updated automatically. To avoid that, you need to set the `style` of every buffer to the newly-generated one, which is not exactly elegant. Not sure it belongs to this patch.

**Comment by aartaka** [comment]
Also, I've forgotten to check the color contrast (see #2115) of this theme when we were thinking of it. Here's the results I've got, they are not perfect:

```
text+primary = AA
background+primary = AA for graphics and large text only
text+secondary = fail
background+tertiary = AA
text+quaternary = AAA

accent+text = fail
accent+background = AA
accent+primary = fail
accent+secondary = fail
accent+tertiary = fail
accent+quaternary = AA
```

Accent color is fine, but `background+primary` (probably the most frequent interface color combination we have) and `text+secondary` we should fix, I guess. @pdelfino, are you interested in experimenting with the theme to make it more contrasting, or should I do that?

**Comment by pdelfino** [comment]
> Accent color is fine, but `background+primary` (probably the most frequent interface color combination we have) and `text+secondary` we should fix, I guess. @pdelfino, are you interested in experimenting with the theme to make it more contrasting, or should I do that?

I will try that on Monday! Thanks for all the feedback :)

**Comment by aartaka** [comment]
The simple fix would be to swap the `primary` and the `tertiary` colors -- current `tertiary` color has a nice contrast with both `text` and `background` colors.

File: source/user-interface.lisp (hunk 1 above)
Comment (pdelfino) [reply source/user-interface.lisp:30]: An accident with the command `upcase-region` :blush: . I will fix it!

**Comment by pdelfino** [comment]
> Regarding commands

Commands already removed. 

> The simple fix would be to swap the `primary` and the `tertiary` colors -- current `tertiary` color has a nice contrast with both `text` and `background` colors.

Contrast is fixed! :+1: 

> Regarding color snippets

I tried an implementation that generates these results:

Screenshot from 2022-02-24 19-21-39
Screenshot from 2022-02-24 19-21-56

Since the results look kind of ugly,  I am afraid that I might have misunderstood something. Not sure if this is what you had in mind. Did I miss something?

**Comment by aartaka** [comment]
What I though about is rather something along the lines of
``` diff
modified   source/help.lisp
@@ -460,19 +460,33 @@ CLASS is a class symbol."
                                               (nyxt/emacs-mode:emacs-mode :activate nil)
                                               (nyxt/vi-mode:vi-normal-mode :activate t))))
                      "Use vi"))
-        (:h2 "Theme style")
-        (:p (:button :class "button"
-                     :onclick (ps:ps (nyxt/ps:lisp-eval
-                                      `(configure-slot 'theme 'browser
-                                                       :value 'theme::+light-theme+)))
-                     (:span :style (format nil "background-color: ~a" (theme:primary-color theme::+light-theme+))
-                            "Use default (Light theme)")))
-        (:p (:button :class "button"
-                     :onclick (ps:ps (nyxt/ps:lisp-eval
-                                      `(configure-slot 'theme 'browser
-                                                       :value 'theme::+dark-theme+)))
-                     (:span :style (format nil "background-color: ~a" (theme:primary-color theme::+dark-theme+))
-                            "Use Dark theme")))
+        (flet ((generate-colors (theme-symbol text)
+                 (spinneret:with-html-string
+                   (:p (:button :class "button"
+                                :style (format nil "background-color: ~a; color: ~a"
+                                               (theme:primary-color (symbol-value theme-symbol))
+                                               (theme:background-color (symbol-value theme-symbol)))
+                                :onclick (ps:ps (nyxt/ps:lisp-eval
+                                                 `(configure-slot 'theme 'browser
+                                                                  :value ',theme-symbol)))
+                                text))
+                   (:p "Colors:")
+                   (:dl
+                    (loop for (name color text-color) in '(("Text" theme:text-color theme:background-color)
+                                                           ("Accent" theme:accent-color theme:background-color)
+                                                           ("Primary" theme:primary-color theme:background-color)
+                                                           ("Secondary" theme:secondary-color theme:background-color)
+                                                           ("Tertiary" theme:tertiary-color theme:text-color)
+                                                           ("Quaternary" theme:quaternary-color theme:text-color)
+                                                           ("Background" theme:background-color theme:text-color))
+                          collect (:dt name ": ")
+                          collect (:dd (:span :style (format nil "background-color: ~a; color: ~a; border-radius: 0.2em"
+                                                             (slot-value (symbol-value theme-symbol) color)
+                                                             (slot-value (symbol-value theme-symbol) text-color))
+                                              (slot-value (symbol-value theme-symbol) color))))))))
+          (:h2 "Theme style")
+          (:raw (generate-colors 'theme::+light-theme+ "Use default (Light theme)"))
+          (:raw (generate-colors 'theme::+dark-theme+ "Use Dark theme")))
         (:h2 "Default new buffer URL")
         (:button :class "button"
                  :onclick (ps:ps (nyxt/ps:lisp-eval
```

This way, all the colors of the theme are visible and one is aware of what they choose, not only by primary color, but by the full list :)

Alternatively, define a `value->html` method for `theme:theme` and list colors beautifully here! This way, people inspecting themes can see the colors too :D

**Comment by aartaka** [comment]
And remember: themes provide their own sets of colors, sou you better set both `background-color` and `color` with the ones theme provides. Otherwise colors might get diluted by the current theme :)

**Comment by pdelfino** [comment]
Thanks, @aartaka! Your suggestion was totally out of sight for me. I think this is good to go now. What do you think?

File: source/help.lisp [hunk 4]
```diff
@@ -431,53 +431,81 @@ CLASS is a class symbol."
...
-                                `(nyxt::configure-slot 'default-new-buffer-url 'browser :type 'STRING)))
-               "Set default new buffer URL")
-      (:h2 "Default zoom ratio")
-      (:button :class "button"
-               :onclick (ps:ps (nyxt/ps:lisp-eval
-                                `(nyxt::configure-slot 'current-zoom-ratio 'buffer)))
-               "Set default zoom ratio")
-      (:h2 "Disable compositing")
-      (:p "On some systems, compositing can cause issues with rendering. If you
+      (theme:with-theme (nyxt:theme nyxt:*browser*)
```
Comment (aartaka) [inline source/help.lisp:434]: Do you use it anywhere, actually? `theme:with-theme` binds `theme:text`, `theme:accent`  etc. so you should only use it if you have to generate CSS with colors injected.

**Comment by aartaka** [comment]
Just tested it and it's a wonderful theme! Thanks for taking the time to implement it, Pedro! Merging right now :D

**Comment by pdelfino** [comment]
> Just tested it and it's a wonderful theme! Thanks for taking the time to implement it, Pedro! Merging right now :D

Thank you, Artyom. It was mainly your contribution :)
I learned a bit about themes due to you. Thanks!

**Pull request merged**